### PR TITLE
HADOOP-16853. ITestS3GuardOutOfBandOperations failing on versioned S3 buckets

### PR DESCRIPTION
**Pull request by steveloughran**
Tested s3 ireland with/without change.detection.source = versionid

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 29s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  27m  0s |  trunk passed  |
| +1 :green_heart: |  compile  |   1m  1s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 34s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 58s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  21m 59s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 32s |  trunk passed  |
| +0 :ok: |  spotbugs  |   1m 26s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   1m 23s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 42s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 42s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 42s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 29s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 48s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  20m 47s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 30s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m 50s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 19s |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 45s |  The patch does not generate ASF License warnings.  |
|  |   |  84m  3s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.4 Server=19.03.4 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1840 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux bef3526a60dc 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e637797 |
| Default Java | 1.8.0_242 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/1/testReport/ |
| Max. process+thread count | 393 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
@bgaborg @mukund-thakur @liuml07  -anyone fancy a quick look @ this? 
tested the codepaths with s3guard on and the bucket with/without version checking. (joy, a test which depends on S3 bucket settings for full coverage)

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java [hunk 1]
```diff
@@ -969,16 +970,42 @@ private void deleteFileInListing()
...
+      if (isVersionedChangeDetection() && status.getVersionId() != null) {
+        // when the status entry has a version ID, then that may be used
+        // when opening the file on what is clearly a versioned store.
+        int length = text.length();
+        byte[] bytes = readOpenFileAPI(guardedFs, testFilePath, length, null);
+        Assertions.assertThat(toChar(bytes))
+            .describedAs("openFile(%s)", testFilePath)
+            .isEqualTo(text);
+        // reading the rawFS with status will also work.
+        bytes = readOpenFileAPI(rawFS, testFilePath, length, status);
```
Comment (liuml07) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java]: `bytes = readOpenFileAPI(rawFS, testFilePath, length, null);` should still fail right? Do you think we can also that in this `if` clause?

**Comment by steveloughran** [comment]
> `bytes = readOpenFileAPI(rawFS, testFilePath, length, null);` should still fail right? Do you think we can also that in this `if` clause? 

yeah, you are right. I'll include that. It should be gone from the rawfs

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java (hunk 1 above)
Comment (bgaborg) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java]: `bytes = readOpenFileAPI(rawFS, testFilePath, length, null);` won't fail (tested). Why should it fail? The`FileStatus` is just an additional parameter for the`FutureDataInputStreamBuilder`.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java [hunk 2]
```diff
@@ -969,16 +970,42 @@ private void deleteFileInListing()
...
 
       // File status will be still readable from s3guard
-      FileStatus status = guardedFs.getFileStatus(testFilePath);
+      S3AFileStatus status = (S3AFileStatus)
+          guardedFs.getFileStatus(testFilePath);
       LOG.info("authoritative: {} status: {}", allowAuthoritative, status);
-      expectExceptionWhenReading(testFilePath, text);
-      expectExceptionWhenReadingOpenFileAPI(testFilePath, text, null);
-      expectExceptionWhenReadingOpenFileAPI(testFilePath, text, status);
+      if (isVersionedChangeDetection() && status.getVersionId() != null) {
```
Comment (bgaborg) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java]: I would rather add this to the test as a new private method.

Comment (liuml07) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java]: This is very specific to this test, which is not too long to me; and we don't need to pass some parameters to the new helper method. I’m happy it sits here or a new method.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java (hunk 1 above)
Comment (liuml07) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java]: > bytes = readOpenFileAPI(rawFS, testFilePath, length, null);

The file was deleted and I was expecting FNF exception here. I have not tested here with versioned  bucket.

Comment (steveloughran) [reply hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java]: `bytes = readOpenFileAPI(rawFS, testFilePath, length, null)` should fail; no status coming in, will add the check

**Comment by steveloughran** [comment]
Tested s3 ireland, versioned bucket and toggled versionid change detection across a pair of runs

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java [hunk 3]
```diff
@@ -1032,8 +1034,26 @@ private void expectExceptionWhenReading(Path testFilePath, String text)
...
+   * @param fs filesystem
+   * @param testFilePath path of the test file
+   * @param text the context in the file.
+   * @param status optional status for the withFileStatus operation.
+   * @throws Exception failure other than the FNFE
+   */
+  private void expectExceptionWhenReadingOpenFileAPI(
+      final S3AFileSystem fs,
+      final Path testFilePath
+      , final String text,
```
Comment (liuml07) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/ITestS3GuardOutOfBandOperations.java]: nit: the `,` here goes to the end of last line... :)

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  30m  8s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  21m 49s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 33s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 23s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 37s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 40s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 26s |  trunk passed  |
| +0 :ok: |  spotbugs  |   1m  0s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 58s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 32s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 26s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 26s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 17s |  hadoop-tools/hadoop-aws: The patch generated 1 new + 2 unchanged - 0 fixed = 3 total (was 2)  |
| +1 :green_heart: |  mvnsite  |   0m 31s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  16m 13s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 24s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m  4s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 15s |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 35s |  The patch does not generate ASF License warnings.  |
|  |   |  94m 16s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.5 Server=19.03.5 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1840 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 8853fa34b0b9 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a562942 |
| Default Java | 1.8.0_242 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/2/testReport/ |
| Max. process+thread count | 423 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by liuml07** [comment]
> The whole deleteFileInListing method is just a redirection from testListingDelete.

Yeah, this is also a good catch yet not addressed in code.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 30s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  20m  6s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 37s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 25s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 38s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  15m 10s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 27s |  trunk passed  |
| +0 :ok: |  spotbugs  |   1m  3s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   1m  0s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 35s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 27s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 27s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 18s |  hadoop-tools/hadoop-aws: The patch generated 1 new + 2 unchanged - 0 fixed = 3 total (was 2)  |
| +1 :green_heart: |  mvnsite  |   0m 33s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  13m 51s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m 16s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 34s |  hadoop-aws in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 32s |  The patch does not generate ASF License warnings.  |
|  |   |  59m 49s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.6 Server=19.03.6 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1840 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 7ed360530477 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 42dfd27 |
| Default Java | 1.8.0_242 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/3/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/3/testReport/ |
| Max. process+thread count | 414 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1840/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
thanks

**Pull request merged**